075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+4. From the **(N+1)th** to the **(M-1)th** reconciliation (rebalancing is going on, checks when it's done):
+     * the `podSet()` does nothing because no new brokers are added this time. 
+     * the `autoRebalancing()` has to check the status of the rebalancing by looking at the corresponding `KafkaRebalance` custom resource status which will move through `ProposalPending`, `ProposalReady` to `Rebalancing` (we are assuming auto-approval). The rebalancing status is reflected into the `Kafka.status.cruiseControl.autoRebalance.state` field.
+5. On the **Mth** reconciliation (rebalancing id done):
+     * the `podSet()` does nothing because no new brokers are added this time.
+     * the `autoRebalancing()` detects that the `KafkaRebalance` custom resource status reports the `Ready` state (so the rebalancing is completed).
+
+During the above flow, errors can happen after the scaling up:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state which is reflected into the `Kafka.status.cruiseControl.autoRebalance.state` field.
+
+It's the user who should deal with this scenario:
+
+* Checking the reason why the `KafkaRebalance` is in the `NotReady` state and :
+  * deleting the custom resource if they are not interested in the rebalancing anymore. Scaling up is done already.
+  * setting the `strimzi.io/rebalance: refresh` annotation on the `KafkaRebalance` custom resource to try the rebalancing again.
+
+### Auto-rebalancing on scaling down
+
+The following flow is triggered when the user decreases the `replicas` field within a `Kafka` or `KafkaNodePool` custom resource in order to scale down the cluster by removing brokers.
+
+1. On the **1st** reconciliation (scaling down is skipped, rebalancing started):
+     * the check on scaling down the cluster fails, but it doesn't have to revert the `replicas` back because the `Kafka.spec.cruiseControl.autoRebalance` is provided. So the reconciliation doesn't fail.
+     * the `scaleDown()` method skips doing the scale down because a rebalancing is needed first (and the check failed).
+     * the `autoreRalancing()` method detects that a scaling down was requested because of `KafkaCluster.removedNodes()` returning a non empty list with the brokers' IDs to be removed from the cluster. Because the `Kafka.status.cruiseControl.autoRebalance.state` doesn't exist, it's a new operation so a new `KafkaRebalance` custom resource is created and the `Kafka.status.cruiseControl.autoRebalance.state` is updated with the corresponding state (should be `ProposalPending`) and the `remove-brokers` with the brokers' IDs.
```

Comment:
The removed nodes would not give you this information. The scale-down relies on the scale-down prevention, but that also reverts the `removedNodes()`. You would need to do some special collection from the KafkaClusterCreator class before the change is reverted. I also commented earlier on the option to disable the scale-down prevention etc.

Reply:
I explained in a previous comment and also on the first bullet point here that the change on replicas should not be reverted because an auto-rebalance is requested.
So the check fails (nodes are hosting partition), it would fail and revert but in this case because the auto-rebalancing is set on scale-down, it doesn't fail and doesn't revert allowing the reconciliation to proceed and `removeNodes()` providing the needed info for the rebalancing.

Reply:
So, the pods will be gone before you rebalance it? Also, will you delete the whole logic for reverting the scale-down that was implemented exactly for this purpose?

Reply:
> So, the pods will be gone before you rebalance it?

The detailed process says ...

`the `scaleDown()` method skips doing the scale down because a rebalancing is needed first (and the check failed)`

So I would tweak the check by not reverting the replicas, not failing the reconciliation but the `scaleDown()` is aware of the auto-rebalancing so it should skip scale down.

> Also, will you delete the whole logic for reverting the scale-down that was implemented exactly for this purpose?

Nope. Just skipping the revert when autorebalancing is set.

Reply:
Yes I will set up call for this but, reviewing the code again and looking at your comment ...

* Having the rebalancing as a sync operation blocking the reconciliation is a no go for me. Having the call to CC directly withing the KAO is still a no go (see alternative 3) as per our previous call about avoiding too tight relation between KAO and CC rebalancing.
* I got your point. I can just let the operator doing the reverting but what I would need is my own `removedNodes` list (because it's emptied on reverting) for the autorebalancing (as you suggested in your previous comment).

I guess the things could work this way ...

1. Of first reconcile, the scale-down is prevented and reverted (in memory) but the `Kafka` or `KafkaNodePool` CR still has the scaled down replicas (which is important for the next reconcile). The auto-rebalancing can be triggered and reconciliation ends.
2. On next N reconciliations, the scale-down is still checked (because CR has the scaled down replicas) and it still fails if brokers host partition (auto rebalancing didn't end yet). The auto-rebalancing part checks and update the status of moving partitions.
3. At some point, at reconciliation M, the scale-down check should not fail anymore because auto-rebalancing ends so the scale-down happens.

Does it make more sense?

Reply:
> If it was as simple as skipping the scaleDown() step, we would have spent a lot less time on it with Shubham. The scaled-down affects many other things such as secrets, services, routes, ingresses etc. You might also hit duplicate node IDs if you don't properly handle it everywhere etc.

Given this, would it be so bad if the rebalance call was blocking? Especially, if autorebalancing were to be an opt-in feature? It sounds like there a lot of components that would need to wait for a successful rebalance before proceeding with a scale-down anyway.

If not,  I am sure we could cover all the cases even if it is complicated to do. Maybe it is worth hacking this part of the feature upstream (if not done already) to better gauge the obstacles of this direction and see if we can  solve them in a reasonable way. Happy to help hack this myself if needed!

Reply:
I know it's going to be complicated but I think we should stick with having rebalancing running across reconciliations and not blocking. This is the way the operator works in general even for long running operations. Having the inner loop blocking the reconciliation in the past was a mistake imho.